src/hotspot/share/gc/shared/gcArguments.cpp

```diff
@@ -138,6 +141,15 @@ void GCArguments::initialize_heap_flags_and_sizes() {
     vm_exit_during_initialization("Too small minimum heap");
   }
 
+#if defined(LINUX)
+  if (OSContainer::is_containerized() && FLAG_IS_CMDLINE(InitialHeapSize)) {
+    jlong limit = OSContainer::memory_and_swap_limit_in_bytes();
+    if ((limit > 0) && (InitialHeapSize >= (julong) limit)) {
+      vm_exit_during_initialization("Initial heap size set to a larger value than the container memory & swap limit");
+    }
```

Comment:
Since you are testing `InitialHeapSize >= (julong) limit)`, the error message should be:

> Initial heap size should be smaller than the container memory & swap limit

The JVM will use memory outside of the heap (for example malloc, or code cache), so even if you set the `InitialHeapSize` to below the memory limit, the JVM might still get killed by OOM unpredictably.

This PR catches obvious problems when `-Xms` is set way too large. However, I think we need to consider the larger issue of OOM killer when using Java in containers. I filed [JDK-8285277](https://bugs.openjdk.java.net/browse/JDK-8285277)

I've started a discussion on [hotspot-dev](https://mail.openjdk.java.net/pipermail/hotspot-dev/2022-April/059463.html) about what approach we should use to handle container memory limits. Let's hold on with this PR until we have a consensus.

Reply:
> Since you are testing `InitialHeapSize >= (julong) limit)`, the error message should be:
> 
> > Initial heap size should be smaller than the container memory & swap limit
> 
> The JVM will use memory outside of the heap (for example malloc, or code cache), so even if you set the `InitialHeapSize` to below the memory limit, the JVM might still get killed by OOM unpredictably.
> 
> This PR catches obvious problems when `-Xms` is set way too large. However, I think we need to consider the larger issue of OOM killer when using Java in containers. I filed [JDK-8285277](https://bugs.openjdk.java.net/browse/JDK-8285277)
> 
> I've started a discussion on [hotspot-dev](https://mail.openjdk.java.net/pipermail/hotspot-dev/2022-April/059463.html) about what approach we should use to handle container memory limits. Let's hold on with this PR until we have a consensus.

Thanks
The code I want to achieve here is similar to the physical machine situation, with only Xms memory guaranteed